[Ho Zong Han] iP

sharpstorm opened the pull request:
# PR Summary
This PR overhauls the skeleton code in the upstream repo to include more functionality and turn it into a useful task tracker 📈.
See [Pull Request Page](https://github.com/sharpstorm/ip/pulls?q=is%3Apr+is%3Aclosed) for full details.

## Naming Change
This PR *changes* the name of the bot from ~Duke~ to **Megumin**.

## Features Added (By Level) ✅

1. Greeting and Exit messages
2. Task creation and Task listing
3. Task Completion tracking and marking
4. Different Task types for Todo, Deadline and Event tasks
5. Error handling and feedback messages
6. Task deletion
7. Saving and Loading of tasks from a persistent data file
8. DateTime-awareness for Deadline and Event tasks
9. Searching of keywords in task list

## Software Design
The code in this PR also considers good software engineering practices, including:

- Object-Oriented Programming centric class-level design
- Packaging and use of accessibility modifiers to restrict outward-facing endpoints
- Unit tests using JUnit
- Inclusion of JavaDocs for all non-private members and methods
- Adherence to CS2103 Coding standard & Google Java style guide

## Main Entry Point
The main entry point for the application can be found in the `Duke` class
```Java
public static void main(String[] args) {
    new Duke().init().run();
}
```

# Pre-Release Binary
A binary is available for feature level 8, available from [here](https://github.com/sharpstorm/ip/releases/tag/A-Jar).
Run the java file using the following command: `java -jar Megumin.jar`
> Note that the application requires read and write access to the directory it is in.

# PR Revision
📆 This PR is updated to reflect the following weeks:

- [x] Week 1
- [x] Week 2
- [x] Week 3
- [ ] Week 4
- [ ] Week 5
- [ ] Week 6
- [ ] Week 7
- [ ] Week 8
- [ ] Week 9
- [ ] Week 10
- [ ] Week 11
- [ ] Week 12
- [ ] Week 13

Commit:
Add Gradle support

sharpstorm committed:
Add Bot Welcome Greeting

sharpstorm committed:
Add User Input Functionality

sharpstorm committed:
Add Functional Interface for block printing

sharpstorm committed:
Add exit command

sharpstorm committed:
Merge pull request #1 from sharpstorm/branch-Level-1

Update feature set to Level-1

Adds the following features:
- Initialization and Greeting
- Command Echoing
- Exit Command

sharpstorm committed:
Add Task and TaskStore data structures

sharpstorm committed:
Add Task Creation and Listing

sharpstorm committed:
Update user input prompt to reflect task creation

sharpstorm committed:
Merge pull request #2 from sharpstorm/branch-Level-2

Update feature set to Level-2

Adds the following features:
- Task Creation
- Task Listing

sharpstorm committed:
Add completion marker to Task data structure

sharpstorm committed:
Update task listing to include done indicator

sharpstorm committed:
Add Task marking and unmarking functionality

sharpstorm committed:
Update marking to show no action messages

sharpstorm committed:
Merge pull request #3 from sharpstorm/branch-Level-3

Update feature set to Level-3

Adds the following features:
- Task Completion Tracking
- Task Completion Marking

sharpstorm committed:
Add new Task types and data fields

Task class: make Task an abstract class

The task class is currently a concrete class, while it is should be
abstract under the new modelling.

Let's,
* Make Task an abstract class
* Remove the addTask method with String argument in TaskStore
* Replace Duke driver Task creation with placeholder Todo type

sharpstorm committed:
Tighten command matching to exact word

Change command matching logic.
The previous command matching used startWith as the matching
mechanism. This could resulted in unintentionally matching
command words with extra trailing characters.

Let's
* Change matching logic to be exact, in other words, use equals
* Streamline command processing flow

sharpstorm committed:
Add Task type printing

sharpstorm committed:
Optimise command processing flow

sharpstorm committed:
Add Event and Deadline creation

sharpstorm committed:
Merge pull request #4 from sharpstorm/branch-Level-4

Update feature set to Level-4

Adds the following features:
- Distinct Todo, Deadline and Event Task types
- Creation of different Task types

Fixes the following:
- Command matching was not exact. Changed command parsing logic
- Streamline command processing flow

sharpstorm committed:
Update test cases and test script

sharpstorm committed:
Merge pull request #5 from sharpstorm/text-ui-test

Update test cases and test script

sharpstorm committed:
Add exception-based error handling

sharpstorm committed:
Update test cases for new errors

sharpstorm committed:
Merge pull request #6 from sharpstorm/branch-Level-5

Update feature set to Level-5

Adds the following features:
- Exception-based error handling
- More specific errors messages

Updates the following:
- Test cases

sharpstorm committed:
Add delete task functionality

sharpstorm committed:
Update test cases for testing delete

sharpstorm committed:
Clean up code formatting

sharpstorm committed:
Merge pull request #7 from sharpstorm/branch-Level-6

Update feature set to Level-6

Add the following features:
- Task Deletion

Update the following:
- Test cases for task deletion

sharpstorm committed:
Add Task serialization and inflation

sharpstorm committed:
Fix Stackoverflow error in TaskType

sharpstorm committed:
Add saving and loading of Task list

sharpstorm committed:
Refactor Task saving to Observer pattern

sharpstorm committed:
Fix leaking file streams in serializers

sharpstorm committed:
Add final modifier to TaskType private variables

sharpstorm committed:
Add parsing of date object for dated Tasks

sharpstorm committed:
Beautify date printing format

sharpstorm committed:
Add upcoming and schedule commands

sharpstorm committed:
Add between command

sharpstorm committed:
Merge PR #8 from branch-Level-7

Add persistence to task list

Task list is lost upon exit of application.

A persistent task list is more practical as a user might want to update
the task list at different points in time, across multiple application
runs.

Let's
- Add saving of the task list to a persistent storage
- Add loading of the task list from a persistent storage

sharpstorm committed:
Merge PR #9 from branch-Level-8

Update feature set to Level-8

Add the following features:
- Parser for date strings
- Upcoming command for listing events within x days
- Between command for listing events within period
- Schedule command for listing events on a day

sharpstorm committed:
Refactor TaskStore to TaskList

sharpstorm committed:
Separate Serializer from file system logic

sharpstorm committed:
Refactor TaskStoreSerializer to TaskListSerializer

sharpstorm committed:
Extract command parsing logic to separate class

Extract command parsing to separate class

Command processing is all performed in the Duke class.
This results in a very long file, with no clear separation of
responsibilities. It also makes the file messy with UI and parsing
logic.

Let's
- Extract command parsing to the Parser class
- Hide command processing logic from the Duke class

sharpstorm committed:
Refactor UI related functions to a separate class

See commit 8a43906 for reasoning.

sharpstorm committed:
Add error message for IO exceptions

sharpstorm committed:
Extract commands to individual classes

Extract commands to multiple classes

All command processing logic reside in Parser class. This results in a
long file with many independent methods that can make it difficult to
search for specific processing logic.

This commit changes this by
- Extracting command processing logic to logical groups, each
  encapsulated within a class.
- Creating a hierarchy of Command classes following the strategy
  pattern.
- Changing Parser to follow the factory design pattern.

sharpstorm committed:
Update test script and test cases

sharpstorm committed:
Remove unnecessary imports

sharpstorm committed:
Merge PR #10 from branch-A-MoreOOP

Refactor classes for cleaner structure

Majority of processing logic and data storage are
not logically grouped. This makes the overall code longer
than it needs to be. Searching for relevant chunks of code
can therefore be difficult.

This PR makes the following changes:
- Extract logic into 4 logical subdomains
- Separate command processing logic into smaller files

sharpstorm committed:
Refactor classes into packages

sharpstorm committed:
Refactor command package structure

sharpstorm committed:
Tighten access modifiers across packages

sharpstorm committed:
Refactor entry point Duke into object

sharpstorm committed:
Change Task modification to package only

sharpstorm committed:
Update test script for packages

sharpstorm committed:
Merge PR #11 from branch-A-Packages

Structure classes into java packages

Classes are in a single namespace without any packages.
This makes it difficult to find single and related classes in the
development workflow. Organizing the classes into separate logical
namespaces helps easily identify related classes, and allows for
package-level access protection.

This PR makes the following changes:
- Add logical namespaces in the form of java packages
- Move classes into different packages
- Modify the test script to compile and run after moving classes
  into packages

sharpstorm committed:
Add initial test for Ui

sharpstorm committed:
Refactor visibility of methods in task package

sharpstorm committed:
Add Tests for task package

sharpstorm committed:
Fix formatting bugs in command handlers

sharpstorm committed:
Add command handler test cases

sharpstorm committed:
Add exit check for Exit command handler

sharpstorm committed:
Add test cases for Ui

sharpstorm committed:
Merge PR #12 from branch-A-JUnit

Add JUnit testing

Unit testing can help eliminate bugs that are difficult to detect
when using the application. Adding unit tests improves the quality
of the codebase, and leads to better regression detection during
development.

Changes made in this PR are
- Add JUnit testing support
- Add test cases for all public methods in current classes

sharpstorm committed:
Add configuration for building JAR file

sharpstorm committed:
Merge branch 'branch-A-Jar'

sharpstorm committed:
Add find tasks functionality

sharpstorm committed:
Fix find command matching bug

sharpstorm committed:
Update test case for find command

sharpstorm committed:
Rename interfaces to remove leading I

sharpstorm committed:
Format code in command package

sharpstorm committed:
Format code in task package

sharpstorm committed:
Format code in duke package

sharpstorm committed:
Remove extra imports in test scripts

sharpstorm committed:
Format code in test scripts

sharpstorm committed:
Add Javadocs for the duke package

sharpstorm committed:
Add Javadoc for exceptions and interfaces

sharpstorm committed:
Add Javadocs for task package

sharpstorm committed:
Add Javadocs for command package

sharpstorm committed:
Fix small formatting bugs in Javadocs

sharpstorm committed:
Change DukeIOException to comply with standard

sharpstorm committed:
Add Javadoc for find command handler

sharpstorm committed:
Merge PR #13 from branch-A-JavaDoc

Add Javadocs to all non-private methods and members

sharpstorm committed:
Merge PR #14 from branch-A-CodingStandard

Change code formatting to comply with coding standard

sharpstorm committed:
Merge PR #15 from branch-Level-9

Update to feature Level 9

This PR adds the functionality for finding tasks by keywords.

sharpstorm committed:
Fix styling issues in testing scripts

sharpstorm committed:
Fix broken level 9 caused by merge

sharpstorm committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

sharpstorm committed:
Modify gradle project configuration

sharpstorm committed:
Merge PR #16 from sharpstorm/branch-A-Gradle

Add gradle support to project

sharpstorm committed:
Add checkstyle support to project

sharpstorm committed:
Update import ordering to satisfy checkstyle

sharpstorm committed:
Merge PR #17 from branch-A-CheckStyle

Add Checkstyle support

sharpstorm committed:
Add JavaFX gradle dependencies

sharpstorm committed:
Add basic JavaFX ui

sharpstorm committed:
Refactor semantics of Ui hierarchy

sharpstorm committed:
Add resizable ui support

zixin448 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,109 @@
+    private static FileInputStream openDatabaseRead() throws DukeIoException {
+        initDataStore(); // throws DukeIoException
+
+        final File database = Paths.get(DATA_FOLDER_PATH, DB_FILENAME).toFile();
```
Should 'database' be renamed as 'DATABASE' instead?

sharpstorm replied:
Thanks for the reccommendation. Fixed in commit 601a63b.

zixin448 commented on src/main/java/duke/command/BetweenCommand.java:
```diff
@@ -0,0 +1,41 @@
+        if (argParts.length < 2) {
+            throw new DukeIllegalArgumentException("Not in the format <date> and <date>");
+        }
+        final LocalDateTime start = parseDateTime(argParts[0]);
+        final LocalDateTime end = parseDateTime(argParts[1]);
```
Should 'start' and 'end' be renamed as 'START' and 'END' instead?

sharpstorm replied:
I believe that this is a local final variable rather than a constant (defined by the presence of `static final`), so it should not follow the screaming snake case.

zixin448 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,104 @@
+     */
+    protected Task parseSelectTask(TaskList taskList) throws DukeIllegalArgumentException {
+        int taskIndex = this.parseTaskNumber();
+        Task task = taskList.getTaskByIndex(taskIndex);
```
I like the way you name your methods! Helps to understand what the method does.

zixin448 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,128 @@
+     */
+    public static Task inflate(TaskType type, DataInputStream dIn) throws IOException {
+        Task task;
+        switch (type) {
+        case TODO:
+            task = new Todo();
+            break;
+        case EVENT:
+            task = new Event();
+            break;
+        case DEADLINE:
+            task = new Deadline();
+            break;
+        default:
+            return null;
+        }
```
Good job for correct indentation for switch-case statements!

sharpstorm committed:
Add Javadocs for newly created methods

sharpstorm committed:
Clean-up imports

sharpstorm committed:
Remove unused variable and extra throws clause

sharpstorm committed:
Add mockito-based ui testing

sharpstorm committed:
Fix up code to comply with coding standard

sharpstorm committed:
Merge PR #18 from branch-Level-10

Update to feature Level 10

This removes the Command Line Interface and adds a
Graphical User Interface.

sharpstorm committed:
Change string array to varargs format

sharpstorm committed:
Change naming of database file name constant

yokester98 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,81 @@
+package duke;
+
+import duke.command.Parser;
+import duke.exception.DukeException;
+import duke.exception.DukeIoException;
+import duke.task.TaskList;
+import duke.ui.Ui;
+import javafx.application.Application;
+import javafx.application.Platform;
+import javafx.stage.Stage;
```
```suggestion
import duke.command.Parser;
import duke.exception.DukeException;
import duke.exception.DukeIoException;
import duke.task.TaskList;
import duke.ui.Ui;

import javafx.application.Application;
import javafx.application.Platform;
import javafx.stage.Stage;
```
As there are many imports here, I feel that a line break will make it easier to browse the list.

sharpstorm replied:
Hi, while I agree with you on the neatness, doing so would actually fail CheckStyle based on the given coding standard configuration.

See error message below:
![image](https://user-images.githubusercontent.com/13163634/152351934-0727ff7c-8065-44cf-88f7-47e5c8bd9539.png)

yokester98 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,81 @@
+        this.initializeTaskList();
+    }
+
+    /**
+     * Performs pre-execution initialization of resources required throughout the application lifecycle.
+     * Loads any existing database and attaches observers for saving the database to disk.
+     * @return The current instance of the application being initialized.
+     */
```
```suggestion
    /**
     * Performs pre-execution initialization of resources required throughout the application lifecycle.
     * Loads any existing database and attaches observers for saving the database to disk.
     *
     * @return The current instance of the application being initialized.
     */
```
Good job on all the Javadoc! For this, I think there should be a line break between the description field and return field.

sharpstorm replied:
Thanks for pointing that out. Fixed in commit 734bff7.

yokester98 commented on src/main/java/duke/command/CreateCommand.java:
```diff
@@ -0,0 +1,112 @@
+        // Syntax Check
+        final String[] argParts = this.args.split(" /at ");
+        if (argParts.length < 2) {
+            throw new DukeIllegalArgumentException("Not in the format <Task name> /at <Date>");
```
You have really excellent detailed error messages provided to the user! I'll learn from this thanks! 😃

yokester98 commented on src/main/java/duke/command/MarkCommand.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    @Override
+    public boolean execute(Printable linePrinter, TaskList taskList) throws DukeIllegalArgumentException {
```
Missing Javadoc for this public method here. According to the Java coding standard, public methods are meant to be used by others and the users should not be forced to read the code of the method to understand its exact behavior. The code, even if it is self-explanatory, can only tell the reader how the code works, not what the code is supposed to do.

sharpstorm replied:
This is a overridden abstract method inherited from the `Command` abstract class. Since the module's coding standard does not explicitly mention overridden methods, I am following Google's Style Guide. [Section 7.3.2](https://google.github.io/styleguide/javaguide.html#s7.3-javadoc-where-required) grants an exception for Javadoc to not be present on overridden methods since Javadocs are also implicitly inherited from the superclass.

yokester98 commented on src/main/java/duke/command/Parser.java:
```diff
@@ -0,0 +1,65 @@
+        final String commandLowerCase = commandParts[0].toLowerCase();
+        final String args = command.substring(commandLowerCase.length()).trim();
+
+        switch (commandLowerCase) {
```
Your entire switch statement is so readable due to the static declarations above! I'll start applying this to my switch statements as well 😅

yokester98 reviewed (commented):
Overall very great job! Your code is very well documented and readable. I liked how you separate your classes into multiple packages as well. Keep up the amazing work! 💯

sharpstorm committed:
Fix Javadoc spacing

sharpstorm committed:
Fix gradle classpath error for javadocs task

sharpstorm committed:
Add Java assertions

Add null-check assertions

There are many points in the code where objects declared are assumed
to be not null, that is, the error-handling code is presumed to work
correctly.

This is a dangerous assumption to make, and could result in hidden
bugs and errors arising from inconsistent code states.

Using a trust-but-verify approach, let's add assertions throughout
the application logic to ensure that the error-handling code works
correctly. Doing so flags events when the application state is
inconsistent, allowing bugs to be identified more easily.

Add default case false-assertions

There are several if-else statements that have default cases that are
ignored simply because they should never be reached. No code exists
within these empty cases.

This is also a dangerous assumption to make, as errors in code could
lead to unexpected behaviour.

Let's add false-assertions to these portions in the code to help
identify when these cases are triggered, which indicates that there are
bugs in the code logic.

sharpstorm committed:
Improve code readability in command package

Change magic numbers and strings to constant variables

Magic numbers are used in command processing logic to translate from
a human readable format to a format that is used in the application's
internal state.

Magic numbers and strings can be difficult to decipher if the reader
has no idea what the code logic is trying to do, and negatively
impacts the readability of the codebase.

Let's extract the magic numbers and strings to named constants that
describe the function of the magic strings or constants. This makes
the code easier to read, and friendlier to readers not involved in
development.

Extract common functionality in CreateCommand to methods

The logic in CreateCommand involves 3 entirely separate flows, and are
implemented in 3 separate methods independent from one another.

However, some of the logic, particularly input validation, are
repeated across the 3 flows, resulting in almost identical code
portions.

Let's extract the common functionality into methods so there is only
a single implementation of the validation logic. This makes the logic
more maintainable and reusable.

sharpstorm committed:
Improve code readability in other packages

Break methods into smaller subroutines

The implementation in TaskListSerializer uses a single method to
perform all the serialization and deserialization required for
TaskList objects.

This results in a complex and deeply nested method, with many
subcomponents in the processing logic, which can be difficult to
understand when reading the code.

Let's break the long serialization methods into subroutines to make
the code more readable, and at the same time reduce the complexity
concentrated in a single part in code.

sharpstorm committed:
Clean up imports and tests

sharpstorm committed:
Merge PR #19 from branch-A-CodeQuality

Improves codebase for better readability.

sharpstorm committed:
Merge branch 'master' into branch-A-Assertions

sharpstorm committed:
Merge PR #20 from branch-A-Assertions

Add Java assertions for internal state validation

sharpstorm committed:
Add github actions workflow

sharpstorm committed:
Remove cyclic dependency in Task inflation

sharpstorm committed:
Fix Mockito testing

sharpstorm committed:
Add Mockito testing for serializers

sharpstorm committed:
Fix incorrect error message in schedule command

sharpstorm committed:
Improve feedback from schedule command

sharpstorm committed:
Add time sorting to schedule command

sharpstorm committed:
Merge PR #21 from branch-B-ViewSchedules

Refine view schedule feature.

sharpstorm committed:
Improve upcoming command

sharpstorm committed:
Merge PR #22 from branch-B-Reminders

Refine upcoming command

sharpstorm committed:
Add base for natural date parsing

sharpstorm committed:
Improve date parsing logic

sharpstorm committed:
Add javadocs and positional date parsing

sharpstorm committed:
Add test cases for date parser

sharpstorm committed:
Add day-relative date parsing

sharpstorm committed:
Fix constant declaration for day mappings

sharpstorm committed:
Merge PR #23 from branch-C-NaturalDates

Add Natural Date Parsing.

sharpstorm committed:
Split long test cases into smaller cases

sharpstorm committed:
Comply with stricter Javadoc standards

sharpstorm committed:
Refactor date parsers

sharpstorm committed:
Change look and feel of application

sharpstorm committed:
Add bot name to JavaFX window title

sharpstorm committed:
Fix checkstyle bug

sharpstorm committed:
Add screenshot of Ui

sharpstorm committed:
Add user guide

sharpstorm committed:
Update legacy README
